### DISCUSSION TITLE
Covers looks weirdly pixelated/poorly resized?

### DISCUSSION THREAD

**Discussion by Ralon17** [s0hv/foo_discord_rich#17] [Q&A]
So I got everything set up such that the album art updates and shows on discord properly. But I'm having a minor issue where compared to other images in discord activity (game icons, covers for spotify activity), my images look very low-res, even for the size. Compare these two: https://imgur.com/a/BV8ofbS. The latter I resized to (roughly since it's not a square) 64x64 in paint.net and it looks much more normal compared to what discord is displaying.

My only guess is that because the images are being uploaded in their large resolution glory, discord is resizing/compressing them in a visually suboptimal manner.

Any thoughts/fixes?

**Comment by s0hv** [comment]
The resizing algorithm depends on the browser. Some rather old results can be found here along with the tool used to make the measurements https://entropymine.com/resamplescope/notes/browsers/ If you want to use a specific resizing algorithm, you must incorporate resizing to your upload script, as that has access to the image before upload. Do note that the rendered image size is not always the same on discord.

**Reply by Azuriye** [reply]
Does Discord use small image key or large image key? I don't understand how the resizing algorithm works... Does it not resize if the images follow a certain resolution or file size? maybe 64x64 resolution or 512x512?

**Reply by s0hv** [reply]
@Azuriye The resizing algorithm is browser dependent and afaik resizing is done on the machine that renders the image. The images are resized (it might also crop the images but I don't remember) if their size differs from the rendered size. And the render size is not constant.

**Reply by Azuriye** [reply]
@s0hv What I've noticed is that there's a RPC implementation for Plex which is again a media player and someone made PlexampRPC on github and surprisingly the images don't get too pixelated. I was under the impression that maybe discord was purposely compressing my images until I tried out PlexampRPC.

For reference: https://github.com/Dyvinia/PlexampRPC

**Reply by s0hv** [reply]
@Azuriye The Plex implementation resizes the images to 128x128 resolution by default. There are two reasons for why I have not done any resizing. First is that discord recommends high resolution images (1024x1024) [[developer docs link](https://discord.com/developers/docs/rich-presence/best-practices#have-interesting-expressive-art)]. Secondly, spotify covers are also high resolution covers with all I've seen being 640x640 resolution. Then, resized in the client to either 60x60 or 90x90. Don't know if other sizes are possible.

If you want to rescale the images yourself that is possible if you make your own script for image uploads.

**Reply by Azuriye** [reply]
@s0hv When resizing images does it do any sort of anti-aliasing or just straight-up downsizing it to 128x128? I assume if the image resolution is less than 128x128 it just sends the image as it is? What about portrait images do they convert that to 128x128 as well or just send them directly to Discord?

Sorta confused about how different RPC connectors for music player work, like do they upsize if they meet a certain resolution threshold or downsize?

**Reply by s0hv** [reply]
@Azuriye As can be seen on [this line](https://github.com/Dyvinia/PlexampRPC/blob/3ae8eed3792a9b13e94d75a3e83598b4465bd939/Windows/MainWindow.xaml.cs#L326), the image resizing is done in plex. I don't know what type of rescaling algorithm is in use there and how exactly does it work. It just reuploads the plex rescaled image to another image host and that url is passed to discord.

As for discord, it depends on the browser, but the image is resized to the required size be it up or downscaling. The desktop application runs on chromium and chromium source code tells us that the the algorithm choice is very situational depending on multiple factors such as hardware [[source](https://source.chromium.org/chromium/chromium/src/+/main:skia/ext/image_operations.h;l=24?ss=chromium)]. I also checked that the longer side of the image is cropped so the image is always a square.

**Reply by Azuriye** [reply]
@s0hv How about DiscordBee? Similar to fb2k but for MusicBee, I honestly don't know how to avoid the pixelated issue but what I've noticed is that images which aren't 1:1 ratio tend to get pixelated alot.

**Reply by s0hv** [reply]
@Azuriye It seems to use [bicubic interpolation](https://github.com/sll552/DiscordBee/blob/8f5b7a137ecbed7fa817ccf8e25f66760c38d6c2/DiscordTools/Assets/Uploader/ResizingUploader.cs#L51). It also resizes the images to a resolution of 1024x1024.

The more detailed the image is the worse it looks downscaled as much detail is lost. Also, I don't think the aspect ratio affects the result.

**Reply by s0hv** [reply]
@Azuriye I'd like to make on correction based on my observations. It seems that the discord app does a bad quality rescale when opening the profile and going to the activity tab. The upper picture is taken from chrome and the below from the app on windows.
![image](https://github.com/s0hv/foo_discord_rich/assets/17063236/77d31792-1ad0-4586-b056-f59ab4b9e2c0)
![image](https://github.com/s0hv/foo_discord_rich/assets/17063236/bfe21ce7-2d1d-40e0-9fdb-fc342e503c73)

However, the smaller image shown when just clicking a username does seem to be the same between the app and browser versions.

**Reply by Azuriye** [reply]
@s0hv 


```
import requests
import sys
from PIL import Image
from io import BytesIO

url = "https://api.imgur.com/3/image"
client_id = ''

try:
    sys.stdin.reconfigure(encoding='utf-8')
    filename = sys.stdin.read().strip()

    headers = {
        'Authorization': f'Client-ID {client_id}'
    }

    with open(filename, 'rb') as f:
        img = Image.open(f)
        width, height = img.size

        if width > 1024 or height > 1024:
            img.thumbnail((1024, 1024), Image.BICUBIC)

        byte_io = BytesIO()
        img.save(byte_io, format='PNG')
        byte_io.seek(0)

        r = requests.post(url, headers=headers, files={'image': byte_io})

    if not r.ok:
        print(r.text[:1000])
        exit(1)

    print(r.json()['data']['link'], end='')
except Exception as e:
    import traceback
    traceback.print_exc(file=sys.stdout)
    exit(1)
```

I've created a small script using the Pillow module and IO module to avoid saving image files locally but instead, to store them in memory. Any idea what could be improved so that it could match similarly to how DiscordBee resizes?

**Reply by s0hv** [reply]
@Azuriye DiscordBee seems to scale it so that both width and height are at least 1024 pixels, while your implementation only downscales so both sides are 1024 pixels at most (if I understood the documentation and code correctly). If the image is smaller than the requested size img.thumbnail does nothing. I'd recommend you just try out different sizes (such as 256x256, 128x128, or 90x90) and different algorithms and see what you like best.
You can find a comparison table of different resizing algorithms [here](https://pillow.readthedocs.io/en/stable/handbook/concepts.html#filters-comparison-table).

**Reply by Azuriye** [reply]
@s0hv Yeah if the image width or height is greater than 1024 pixels then it downscales it otherwise it'd just send without any resizing.

Your saying DiscordBee scales so that the image is atleast 1024 pixels, so I assume it does both upscaling and downscaling to 1024px?

**Reply by s0hv** [reply]
@Azuriye Yes it does both upscaling and downscaling if I understood the code correctly

**Reply by Azuriye** [reply]
@s0hv According to you do you think it's worth up-scaling images? I'm thinking about whether I should deal with both upscale and downscale because if so then I doubt I need an if condition to apply a thumbnail but instead just apply it to all images, so the only factor over here will be deciding between 1024px or 128px... Also, does it matter if I just choose Lanczos or is it a bit too overkill? Just trying to cut out all factors of images getting heavily pixelated...

EDIT: This is my Python script for now taking care of resizing... I assume it's safe to use thumbnail method to take care of aspect ratio related and just pass the desired resolution and resampling?

```
import requests
import sys
from PIL import Image
from io import BytesIO

url = "https://api.imgur.com/3/image"
client_id = ''

try:
    sys.stdin.reconfigure(encoding='utf-8')
    filename = sys.stdin.read().strip()

    headers = {
        'Authorization': f'Client-ID {client_id}'
    }

    with open(filename, 'rb') as f:
        img = Image.open(f)
        
        img.thumbnail((1024, 1024), resample=Image.BICUBIC)

        byte_io = BytesIO()
        img.save(byte_io, format='PNG')
        byte_io.seek(0)

        r = requests.post(url, headers=headers, files={'image': byte_io})

    if not r.ok:
        print(r.text[:1000])
        exit(1)

    print(r.json()['data']['link'], end='')
except Exception as e:
    import traceback
    traceback.print_exc(file=sys.stdout)
    exit(1)

```

**Reply by s0hv** [reply]
@Azuriye Not really. Personally I don't touch the images at all, but that depends on what kind of cover art you mostly have. I see no reason to upscale as that will at best do nothing and at worst just worsen the quality. As for bicubic vs lanczos, try them both and see which looks better. You probably will not be able to see the difference in your discord status.

**Reply by Azuriye** [reply]
@s0hv It's sort of hard to decide between upscale and downscale should I do some prior checking to see if an image is greater than a certain resolution threshold and then downscale it or just forcefully downscale all images to 128x128 as majority of the images should be greater than 128px and Discord only displays it as 60x60 or 90x90. Also, I've set the Resampling back to Lanczos just to avoid tweaking it a lot.

Secondly, should I use the thumbnail or resize method from Pillow as they both do the same job except one respects the aspect ratio and the other does not?

> The more detailed the image is the worse it looks downscaled as much detail is lost. Also, I don't think the aspect ratio affects the result.

Lastly, if both upscaling and downscaling images lose detail how else could you circumvent it? I feel like upscaling would be much worse than downscaling as you have to guess the missing pixels and it starts becoming a mess unlike downscaling with resampling. You mentioned as well how Spotify only uses 640x640 resolution which basically means they don't apply any resampling there either as well.

**Reply by s0hv** [reply]
@Azuriye As I said I recommend you try out different resolutions on the lower end and compare them. You might even realize that the originally image looks best.

If you use the resize method you must write your own logic for aspect ratio preservation which is not too difficult. It also allows for more control as you decide the exact output resolution.

The best way to avoid any loss in detail is to preserve the original image. Though the image will be resized and cropped in discord.

**Reply by Azuriye** [reply]
@s0hv if you look through DiscordBee's ResizeImage method it doesn't upscale but only downscales the image to 1024px

This line checks the aspectRatio and assigns the smallest ratio essentially stopping any upscale unless I'm interpreting this wrongly...?

```
var ratioX = maxWidth / (double)image.Width;
var ratioY = maxHeight / (double)image.Height;
var ratio = ratioX < ratioY ? ratioX : ratioY;
```

**Reply by s0hv** [reply]
@Azuriye Here's an example with values replaced for an image with a size of 500x600 and max size of 1024x1024
 ```cs
ratioX = 1024/500 = 2.048
ratioY = 1024/600 = 1.70667
ratio = 2.048 < 1.70667 => 1.70667

width = 500 *  1.70667 = 853
height = 600 * 1.70667 = 1024
```
As you can see the image is upscaled

**Reply by Azuriye** [reply]
@s0hv I decided to just downscale all images to 128px with Lanczos resample, it does the job fine. 

I tried converting that C# code to Python and use the resize method but when it upscaled it looked too pixelated. 

But it looks like the C# and Python's Pillow thumbnail method are very similar when downscaling.

**Reply by Azuriye** [reply]
> @Azuriye As can be seen on [this line](https://github.com/Dyvinia/PlexampRPC/blob/3ae8eed3792a9b13e94d75a3e83598b4465bd939/Windows/MainWindow.xaml.cs#L326), the image resizing is done in plex. I don't know what type of rescaling algorithm is in use there and how exactly does it work. It just reuploads the plex rescaled image to another image host and that url is passed to discord.

Hello @s0hv, sorry for necro-posting this discussion. But I found out the way Plex parses the image
`https://192-168-0-141:32400/photo/:/transcode?width=360&height=360&minSize=1&upscale=1&url=%2Flibrary%2Fmetadata%2F18274%2Fthumb%2F1705803230%3F`

I've truncated some sensitive information but it looks like it's downscaling images to 360x360 but not sure what minSize and upscale does... any clue?

**Reply by s0hv** [reply]
According to [this documentation](https://python-plexapi.readthedocs.io/en/latest/modules/server.html#plexapi.server.PlexServer.transcodeImage) they are booleans that determine whether to upscale if necessary and whether to maintain the smallest dimension.

**Reply by Azuriye** [reply]
> According to [this documentation](https://python-plexapi.readthedocs.io/en/latest/modules/server.html#plexapi.server.PlexServer.transcodeImage) they are booleans that determine whether to upscale if necessary and whether to maintain the smallest dimension.

Any idea if they could be replicated using python? the functionality for minSize and upscale is closed source so not sure if we could guess this apart from what they do? Also it looks like both the boolean values are always true.

**Reply by s0hv** [reply]
I would guess that minSize determines whether to keep aspect ratio or not, and upscale tells whether to do upscaling when image is smaller than the given resolution.

**Reply by Azuriye** [reply]
Okay good news, after a bit of trial and error, it looks like when minScale is 1, the target dimensions/boundary are 360x360, and this piece of code worked

```
        width, height = img.size
        if width < height:
            new_width = 360
            new_height = round((new_width / width) * height)
        else:
            new_height = 360
            new_width = round((new_height / height) * width)

        img = img.resize((new_width, new_height), resample=Image.LANCZOS)
```

comparing the intrinsic size between Plex's own transcoder and the above implementation matched the dimension. But it looks like discord downscales the image even further down to target boundary 128x128px so replacing the above new_width and new_height with 128 achieves what PlexampRPC displays. I need to figure out regarding whether upScale will be needed or not now.
